### PR TITLE
Feature/watcher notifications

### PR DESCRIPTION
**Pull request by Arkitecht**
- Auto-adds discussion / post creator as a watcher if feature is enabled
- Sends notification email to watchers when a new post is added

### CODE REVIEW

File: src/Listeners/NotifyWatchersPostAdded.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    public function handle(PostAdded $event)
+    {
+        $recipients = $this->getWatchers($event->post);
+
+        if (! $recipients) {
+            return;
+        }
+
+        $recipients->each(fn ($recipient) => Mail::to($recipient)
+            ->send(new PostAddedEmail($event->post)));
```
Comment (AndreasElia) [inline src/Listeners/NotifyWatchersPostAdded.php:40]: This formatting looks meh.

File: src/Services/PostService.php [hunk 2]
```diff
@@ -26,6 +27,11 @@ public function make(Request $request, Discussion $discussion)
 
         $discussion->posts()->save($post);
 
+        if (config('firefly.features.watchers')) {
+            $discussion->watchers()->syncWithoutDetaching([$user->id]);
+            PostAdded::dispatch($post);
```
Comment (AndreasElia) [inline src/Services/PostService.php:32]: New line between these two to make things more readable.

File: tests/Feature/PostTest.php [hunk 3]
```diff
@@ -100,4 +103,75 @@ public function test_content_is_required()
         $response->assertJsonValidationErrors('content');
         $response->assertJson($validJson);
     }
+
+    public function test_watcher_gets_added_when_post_gets_created()
+    {
+        $this->enableWatchersFeature();
+        // Clear all previous posts
```
Comment (AndreasElia) [inline tests/Feature/PostTest.php:110]: New line above the comment.

File: tests/Feature/PostTest.php [hunk 4]
```diff
@@ -100,4 +103,75 @@ public function test_content_is_required()
...
+        $response->assertLocation('forum/d/'.$discussion->uri);
+
+        $this->assertEquals(1, $discussion->watchers()->count());
+        $this->assertEquals($this->getUser()->id, $discussion->watchers()->first()->id);
+    }
+
+    public function test_email_gets_sent_to_watchers_when_post_gets_created()
+    {
+        $this->enableWatchersFeature();
+        Mail::fake();
```
Comment (AndreasElia) [inline tests/Feature/PostTest.php:137]: New line above this line.

**Review by AndreasElia — CHANGES_REQUESTED** [review]

**Comment by AndreasElia** [comment]
Thoughts on including the "features" in the config, but commented out by default?

**Comment by AndreasElia** [comment]
@tomirons do you see any issues with this PR that I may have missed?

**Review by AndreasElia — APPROVED** [review]

File: resources/views/mail/new-post.blade.php [hunk 5]
```diff
@@ -0,0 +1,10 @@
+@component('mail::message')
+{{ __('A new post has been added to the discussion ":discussion"', ['discussion' => $post->discussion->title]) }}
+
+@component('mail::button', ['url' => route('firefly.discussion.show',[$post->discussion->id, $post->discussion->slug])])
```
Comment (AndreasElia) [inline resources/views/mail/new-post.blade.php:4]: Space needed after the 1st first in the route.

**Review by AndreasElia — CHANGES_REQUESTED** [review]

File: src/Events/PostAdded.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * Get the channels the event should broadcast on.
+     *
+     * @return Channel|array
+     */
+    public function broadcastOn()
+    {
+        return new PrivateChannel('channel-name');
```
Comment (tomirons) [inline src/Events/PostAdded.php:42]: This needs a valid channel, no?

**Review by tomirons — CHANGES_REQUESTED** [review]
This is looking really good! I appreciate your contributions on this package! 👍

File: src/Events/PostAdded.php (hunk 6 above)
Comment (Arkitecht) [reply src/Events/PostAdded.php:42]: This is the default scaffolding, when doing an `artisan make:event`

File: src/Services/PostService.php [hunk 7]
```diff
@@ -26,6 +27,12 @@ public function make(Request $request, Discussion $discussion)
 
         $discussion->posts()->save($post);
 
+        if (config('firefly.features.watchers')) {
```
[the remark this answers is not in the document]
Comment (Arkitecht) [reply src/Services/PostService.php:30]: I can work to implement this.

File: src/Events/PostAdded.php (hunk 6 above)
Comment (tomirons) [reply src/Events/PostAdded.php:42]: I'm aware of that. If this event is not supposed to be broadcasted then we can remove that method.

File: src/Services/PostService.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (Arkitecht) [reply src/Services/PostService.php]: I have now implemented this, and update the logic to use the new class

File: src/Events/PostAdded.php (hunk 6 above)
Comment (Arkitecht) [reply src/Events/PostAdded.php]: Apologies, method removed

**Pull request merged**